File: app/src/main/java/zhan/rnadomgirdanimation/refresh/RefreshTask.java
package zhan.rnadomgirdanimation.refresh;

import android.os.Handler;

import java.lang.ref.WeakReference;

public class RefreshTask implements Runnable {

    private WeakReference<RefreshHelper> mWeakRefHelper;
    private WeakReference<Handler> mWeakRefHandler;

    public RefreshTask(RefreshHelper helper, Handler handler) {
        mWeakRefHelper = new WeakReference<>(helper);
        mWeakRefHandler = new WeakReference<>(handler);
    }

    @Override
    public void run() {
        RefreshHelper helper = mWeakRefHelper.get();
        if (helper != null) {
            helper.generateNewData();

            Handler handler = mWeakRefHandler.get();
            if (handler != null && helper.isRunning()) {
                handler.postDelayed(this, helper.getRefreshTime());
            }
        }
    }
}

File: app/src/main/java/zhan/rnadomgirdanimation/refresh/RefreshTask.kt
package zhan.rnadomgirdanimation.refresh

import android.os.Handler

import java.lang.ref.WeakReference

class RefreshTask(helper: RefreshHelper<*>, handler: Handler) : Runnable {

    private val mWeakRefHelper: WeakReference<RefreshHelper<*>> = WeakReference(helper)
    private val mWeakRefHandler: WeakReference<Handler> = WeakReference(handler)

    override fun run() {
        val helper = mWeakRefHelper.get()
        if (helper != null) {
            helper.generateNewData()

            val handler = mWeakRefHandler.get()
            if (handler != null && helper.isRunning) {
                handler.postDelayed(this, helper.refreshTime.toLong())
            }
        }
    }
}
